stress: failed test in cockroach/sql/sql.test: TestPurgeOldLeases

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/82e5c27735be951890f22fc85888a2cf46a13605

Stress build found a failed test:

```
=== RUN   TestPurgeOldLeases
W160522 07:03:16.003824 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160522 07:03:16.005940 storage/engine/rocksdb.go:137  opening in memory rocksdb instance
I160522 07:03:16.007259 server/node.go:401  store store=0:0 ([]=) not bootstrapped
W160522 07:03:16.008908 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160522 07:03:16.009245 storage/replica_command.go:1470  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406639h3m17.008504747s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-05-22 07:03:16.009139725 +0000 UTC]
I160522 07:03:16.009736 server/node.go:327  **** cluster {4b294055-e1e8-42f3-95e4-88714a64f560} has been created
I160522 07:03:16.009758 server/node.go:328  **** add additional nodes by specifying --join=127.0.0.1:36326
W160522 07:03:16.009768 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160522 07:03:16.010496 server/node.go:414  initialized store store=1:1 ([]=): {Capacity:10365558784 Available:7970807808 RangeCount:0}
I160522 07:03:16.010543 server/node.go:302  node ID 1 initialized
I160522 07:03:16.010751 storage/stores.go:287  read 0 node addresses from persistent storage
I160522 07:03:16.010885 server/node.go:535  connecting to gossip network to verify cluster ID...
I160522 07:03:16.011401 server/node.go:556  node connected via gossip and verified as part of cluster {"4b294055-e1e8-42f3-95e4-88714a64f560"}
I160522 07:03:16.011449 server/node.go:355  [node=1] Started node with [[]=] engine(s) and attributes []
I160522 07:03:16.011490 server/server.go:416  starting https server at 127.0.0.1:34563
I160522 07:03:16.011518 server/server.go:417  starting grpc/postgres server at 127.0.0.1:36326
I160522 07:03:16.011926 storage/split_queue.go:100  splitting range=1 [/Min-/Max) at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
I160522 07:03:16.012868 storage/replica_command.go:2011  initiating a split of range=1 [/Min-/Max) at key /Table/11
I160522 07:03:16.020825 server/updates.go:155  No previous updates check time.
I160522 07:03:16.055882 storage/replica_command.go:1470  range 2: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406639h3m17.109982998s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-05-22 07:03:16.055784696 +0000 UTC]
I160522 07:03:16.056019 storage/replica_command.go:2011  initiating a split of range=2 [/Table/11-/Max) at key /Table/12
I160522 07:03:16.068373 storage/replica_command.go:1470  range 3: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406639h3m17.118035764s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-05-22 07:03:16.068331008 +0000 UTC]
I160522 07:03:16.068487 storage/replica_command.go:2011  initiating a split of range=3 [/Table/12-/Max) at key /Table/13
I160522 07:03:16.111656 storage/replica_command.go:1470  range 4: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406639h3m17.161255623s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-05-22 07:03:16.111558814 +0000 UTC]
I160522 07:03:16.111880 storage/replica_command.go:2011  initiating a split of range=4 [/Table/13-/Max) at key /Table/14
I160522 07:03:16.265438 storage/replica_command.go:1470  range 5: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406639h3m17.315065417s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-05-22 07:03:16.265391987 +0000 UTC]
I160522 07:03:16.265570 storage/split_queue.go:100  splitting range=5 [/Table/14-/Max) at keys [/Table/50/0]
I160522 07:03:16.265955 storage/replica_command.go:2011  initiating a split of range=5 [/Table/14-/Max) at key /Table/50
I160522 07:03:16.295464 storage/replica_command.go:1470  range 6: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 406639h3m17.34525492s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-05-22 07:03:16.295422912 +0000 UTC]
I160522 07:03:16.296413 storage/split_queue.go:100  splitting range=6 [/Table/50-/Max) at keys [/Table/51/0]
I160522 07:03:16.296721 storage/replica_command.go:2011  initiating a split of range=6 [/Table/50-/Max) at key /Table/51
--- FAIL: TestPurgeOldLeases (0.31s)
	lease_internal_test.go:153: sql/lease.go:206: DELETE FROM system.lease WHERE (descID, version, nodeID, expiration) = ($1, $2, $3, $4): expected 1 result, found 0
```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
0 runs so far, 0 failures, over 15s
0 runs so far, 0 failures, over 20s
0 runs so far, 0 failures, over 25s
0 runs so far, 0 failures, over 30s
0 runs so far, 0 failures, over 35s
0 runs so far, 0 failures, over 40s
0 runs so far, 0 failures, over 45s
12 runs so far, 0 failures, over 50s
16 runs so far, 0 failures, over 55s
16 runs so far, 0 failures, over 1m0s
16 runs so far, 0 failures, over 1m5s
16 runs so far, 0 failures, over 1m10s
16 runs so far, 0 failures, over 1m15s
16 runs so far, 0 failures, over 1m20s
16 runs so far, 0 failures, over 1m25s
16 runs so far, 0 failures, over 1m30s
16 runs so far, 0 failures, over 1m35s
27 runs so far, 0 failures, over 1m40s
32 runs so far, 0 failures, over 1m45s
32 runs so far, 0 failures, over 1m50s
32 runs so far, 0 failures, over 1m55s
32 runs so far, 0 failures, over 2m0s
32 runs so far, 0 failures, over 2m5s
32 runs so far, 0 failures, over 2m10s
32 runs so far, 0 failures, over 2m15s
32 runs so far, 0 failures, over 2m20s
32 runs so far, 0 failures, over 2m25s
36 runs completed, 1 failures, over 2m26s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/sql/lease.go
+++ b/sql/lease.go
@@ -97,6 +97,8 @@ type LeaseStore struct {
 	db     client.DB
 	clock  *hlc.Clock
 	nodeID uint32
+
+	testingKnobs LeaseStoreTestingKnobs
 }
 
 // jitteredLeaseDuration returns a randomly jittered duration from the interval
@@ -192,6 +194,9 @@ func (s LeaseStore) Acquire(
 // Release a previously acquired table descriptor lease.
 func (s LeaseStore) Release(lease *LeaseState) error {
 	err := s.db.Txn(func(txn *client.Txn) error {
+		if log.V(2) {
+			log.Infof("LeaseStore releasing lease %s", lease)
+		}
 		p := makePlanner()
 		p.txn = txn
 		p.session.User = security.RootUser
@@ -203,10 +208,14 @@ func (s LeaseStore) Release(lease *LeaseState) error {
 			return err
 		}
 		if count != 1 {
-			return util.Errorf("%s: expected 1 result, found %d", deleteLease, count)
+			return util.Errorf("unexpected results while deleting lease %s: "+
+				"expected 1 result, found %d", lease, count)
 		}
 		return nil
 	})
+	if s.testingKnobs.LeaseReleasedEvent != nil {
+		s.testingKnobs.LeaseReleasedEvent(lease, err)
+	}
 	return err
 }
 
@@ -396,7 +405,7 @@ func (l *leaseSet) insert(s *LeaseState) {
 func (l *leaseSet) remove(s *LeaseState) {
 	i, match := l.findIndex(s.Version, s.expiration)
 	if !match {
-		return
+		panic(fmt.Sprintf("can't find lease to remove: %s", s))
 	}
 	l.data = append(l.data[:i], l.data[i+1:]...)
 }
@@ -582,24 +591,16 @@ func (t *tableState) acquireFromStoreLocked(
 // t.mu must be locked.
 // leases must be a not overlap t.active.data, since t.active.data will
 // be changed by this function.
-func (t *tableState) releaseLeasesIfNotActive(
-	leases []*LeaseState, store LeaseStore,
-) error {
-	for _, s := range leases {
-		err := func() error {
-			s.mu.Lock()
-			defer s.mu.Unlock()
-			if s.refcount != 0 {
-				return nil
+func (t *tableState) releaseLeasesIfNotActive(leases []*LeaseState, store LeaseStore) {
+	for _, lease := range leases {
+		func() {
+			lease.mu.Lock()
+			defer lease.mu.Unlock()
+			if lease.refcount == 0 {
+				t.removeLease(lease, store)
 			}
-			t.active.remove(s)
-			return t.releaseNodeLease(s, store)
 		}()
-		if err != nil {
-			return err
-		}
 	}
-	return nil
 }
 
 func (t *tableState) acquireWait() {
@@ -657,8 +658,8 @@ func (t *tableState) release(lease *LeaseState, store LeaseStore) error {
 	}
 	if s.refcount == 0 {
 		if t.deleted {
-			t.active.remove(s)
-			return t.releaseNodeLease(s, store)
+			t.removeLease(s, store)
+			return nil
 		}
 		n := t.active.findNewest(0)
 		if s != n {
@@ -669,8 +670,8 @@ func (t *tableState) release(lease *LeaseState, store LeaseStore) error {
 				// release the lease immediately the transaction will necessarily abort
 				// on the next operation due to not being able to get the lease.
 			}
-			t.active.remove(s)
-			return t.releaseNodeLease(s, store)
+			t.removeLease(s, store)
+			return nil
 		}
 	}
 	return nil
@@ -678,14 +679,16 @@ func (t *tableState) release(lease *LeaseState, store LeaseStore) error {
 
 // t.mu needs to be locked.
 // lease.mu needs to be locked.
-func (t *tableState) releaseNodeLease(lease *LeaseState, store LeaseStore) error {
-	// We're called with mu locked, but need to unlock it while releasing the
-	// lease.
-	t.mu.Unlock()
-	defer t.mu.Lock()
+func (t *tableState) removeLease(lease *LeaseState, store LeaseStore) {
 	lease.released = true
+	t.active.remove(lease)
 	t.tableNameCache.remove(lease)
-	return store.Release(lease)
+	// Release to the store asynchronously, without the tableState lock.
+	go func() {
+		if err := store.Release(lease); err != nil {
+			log.Warningf("Error releasing lease %q: %s", lease, err)
+		}
+	}()
 }
 
 // purgeOldLeases refreshes the leases on a table. Unused leases older than
@@ -723,15 +726,10 @@ func (t *tableState) purgeOldLeases(
 			var toRelease []*LeaseState
 			if deleted {
 				t.deleted = true
-				// If the table has been deleted, all leases are stale.
-				toRelease = append([]*LeaseState(nil), t.active.data...)
-			} else {
-				// Otherwise, all but the lease we just took are stale.
-				toRelease = append([]*LeaseState(nil), t.active.data[:len(t.active.data)-1]...)
 			}
-			if err := t.releaseLeasesIfNotActive(toRelease, store); err != nil {
-				return err
-			}
+			toRelease = append([]*LeaseState(nil), t.active.data...)
+
+			t.releaseLeasesIfNotActive(toRelease, store)
 			return nil
 		}
 		return err
@@ -745,7 +743,18 @@ func (t *tableState) purgeOldLeases(
 	return t.release(lease, store)
 }
 
-// LeaseManagerTestingKnobs contains test affordances.
+// LeaseStoreTestingKnobs contains testing knobs.
+type LeaseStoreTestingKnobs struct {
+	// Called after a lease is removed from the store, with any operation error.
+	LeaseReleasedEvent func(lease *LeaseState, err error)
+}
+
+// ModuleTestingKnobs is part of the base.ModuleTestingKnobs interface.
+func (*LeaseStoreTestingKnobs) ModuleTestingKnobs() {}
+
+var _ base.ModuleTestingKnobs = &LeaseStoreTestingKnobs{}
+
+// LeaseManagerTestingKnobs contains test knobs.
 type LeaseManagerTestingKnobs struct {
 	// A callback called when a gossip update is received, before the leases are
 	// refreshed. Careful when using this to block for too long - you can block
@@ -753,6 +762,8 @@ type LeaseManagerTestingKnobs struct {
 	GossipUpdateEvent func(config.SystemConfig)
 	// A callback called after the leases are refreshed as a result of a gossip update.
 	TestingLeasesRefreshedEvent func(config.SystemConfig)
+
+	LeaseStoreTestingKnobs LeaseStoreTestingKnobs
 }
 
 var _ base.ModuleTestingKnobs = &LeaseManagerTestingKnobs{}
@@ -882,9 +893,10 @@ func NewLeaseManager(
 ) *LeaseManager {
 	lm := &LeaseManager{
 		LeaseStore: LeaseStore{
-			db:     db,
-			clock:  clock,
-			nodeID: nodeID,
+			db:           db,
+			clock:        clock,
+			nodeID:       nodeID,
+			testingKnobs: testingKnobs.LeaseStoreTestingKnobs,
 		},
 		tables:       make(map[sqlbase.ID]*tableState),
 		testingKnobs: testingKnobs,

--- a/sql/lease_internal_test.go
+++ b/sql/lease_internal_test.go
@@ -76,7 +76,6 @@ func TestLeaseSet(t *testing.T) {
 		{newest{2}, "2:4"},
 		{newest{3}, "3:1"},
 		{newest{4}, "<nil>"},
-		{remove{0, 0}, "1:1 2:1 2:2 2:3 2:4 3:1"},
 		{remove{2, 4}, "1:1 2:1 2:2 2:3 3:1"},
 		{remove{3, 1}, "1:1 2:1 2:2 2:3"},
 		{remove{1, 1}, "2:1 2:2 2:3"},

--- a/sql/lease_test.go
+++ b/sql/lease_test.go
@@ -41,21 +41,27 @@ import (
 
 type leaseTest struct {
 	*testing.T
-	server *testServer
-	db     *gosql.DB
-	kvDB   *client.DB
-	nodes  map[uint32]*csql.LeaseManager
+	server                   *testServer
+	db                       *gosql.DB
+	kvDB                     *client.DB
+	nodes                    map[uint32]*csql.LeaseManager
+	leaseManagerTestingKnobs csql.LeaseManagerTestingKnobs
 }
 
 func newLeaseTest(t *testing.T, ctx *server.Context) *leaseTest {
 	s, db, kvDB := setupWithContext(t, ctx)
-	return &leaseTest{
+	leaseTest := &leaseTest{
 		T:      t,
 		server: s,
 		db:     db,
 		kvDB:   kvDB,
 		nodes:  map[uint32]*csql.LeaseManager{},
 	}
+	if ctx.TestingKnobs.SQLLeaseManager != nil {
+		leaseTest.leaseManagerTestingKnobs =
+			*ctx.TestingKnobs.SQLLeaseManager.(*csql.LeaseManagerTestingKnobs)
+	}
+	return leaseTest
 }
 
 func (t *leaseTest) cleanup() {
@@ -118,9 +124,25 @@ func (t *leaseTest) release(nodeID uint32, lease *csql.LeaseState) error {
 	return t.node(nodeID).Release(lease)
 }
 
-func (t *leaseTest) mustRelease(nodeID uint32, lease *csql.LeaseState) {
+// If leaseReleaseWaiter is not nil, it will be used to block until the lease is
+// released from the store. If the lease is not supposed to be released from the
+// store (i.e. it's not expired and it's not for an old descriptor version),
+// this shouldn't be set.
+func (t *leaseTest) mustRelease(
+	nodeID uint32,
+	lease *csql.LeaseState,
+	leaseReleaseWaiter *leaseReleaseWaiter,
+) {
+	if leaseReleaseWaiter != nil {
+		leaseReleaseWaiter.PrepareToWait(lease)
+	}
 	if err := t.release(nodeID, lease); err != nil {
 		t.Fatal(err)
+	}
+	if leaseReleaseWaiter != nil {
+		if err := leaseReleaseWaiter.Wait(); err != nil {
+			t.Fatal(err)
+		}
 	}
 }
 
@@ -142,15 +164,58 @@ func (t *leaseTest) node(nodeID uint32) *csql.LeaseManager {
 	mgr := t.nodes[nodeID]
 	if mgr == nil {
 		mgr = csql.NewLeaseManager(
-			nodeID, *t.server.DB(), t.server.Clock(), csql.LeaseManagerTestingKnobs{})
+			nodeID, *t.server.DB(),
+			t.server.Clock(),
+			t.leaseManagerTestingKnobs,
+		)
 		t.nodes[nodeID] = mgr
 	}
 	return mgr
 }
 
+// leaseReleaseWaiter can be used to wait for a lease to be removed from the
+// store (leases are removed from the store async w.r.t. LeaseManager
+// operations).
+// To use it, its LeaseReleasedNotification method must be hooked up to
+// LeaseStoreTestingKnobs.LeaseReleasedEvent. Then, every time you want to wait
+// for a lease, call PrepareToWait() before calling LeaseManager.Release(), and
+// then call Wait() for the actual blocking.
+type leaseReleaseWaiter struct {
+	waitingFor *csql.LeaseState
+	doneSem    chan error
+}
+
+func (w *leaseReleaseWaiter) PrepareToWait(lease *csql.LeaseState) {
+	w.waitingFor = lease
+	w.doneSem = make(chan error, 1)
+}
+
+func (w *leaseReleaseWaiter) Wait() error {
+	if w.waitingFor == nil {
+		panic("wasn't told what to wait for")
+	}
+	return <-w.doneSem
+}
+
+func (w *leaseReleaseWaiter) LeaseReleasedNotification(
+	lease *csql.LeaseState, err error,
+) {
+	if lease == w.waitingFor {
+		w.doneSem <- err
+	}
+}
+
 func TestLeaseManager(testingT *testing.T) {
 	defer leaktest.AfterTest(testingT)()
 	ctx := server.MakeTestContext()
+	var releaseWaiter leaseReleaseWaiter
+	ctx.TestingKnobs = base.TestingKnobs{
+		SQLLeaseManager: &csql.LeaseManagerTestingKnobs{
+			LeaseStoreTestingKnobs: csql.LeaseStoreTestingKnobs{
+				LeaseReleasedEvent: releaseWaiter.LeaseReleasedNotification,
+			},
+		},
+	}
 	t := newLeaseTest(testingT, &ctx)
 	defer t.cleanup()
 
@@ -168,7 +233,7 @@ func TestLeaseManager(testingT *testing.T) {
 	if err := t.release(2, l1); err == nil {
 		t.Fatalf("expected error, but found none")
 	}
-	t.mustRelease(1, l1)
+	t.mustRelease(1, l1, nil)
 	t.expectLeases(descID, "/1/1")
 
 	// It is an error to acquire a lease for a specific version that doesn't
@@ -186,18 +251,18 @@ func TestLeaseManager(testingT *testing.T) {
 
 	// When the last local reference on the new version is released we don't
 	// release the node lease.
-	t.mustRelease(1, l3)
+	t.mustRelease(1, l3, nil)
 	t.expectLeases(descID, "/1/1 /2/1")
 
 	// We can still acquire a local reference on the old version since it hasn't
 	// expired.
 	l4 := t.mustAcquire(1, descID, 1)
-	t.mustRelease(1, l4)
+	t.mustRelease(1, l4, nil)
 	t.expectLeases(descID, "/1/1 /2/1")
 
 	// When the last local reference on the old version is released the node
 	// lease is also released.
-	t.mustRelease(1, l2)
+	t.mustRelease(1, l2, &releaseWaiter)
 	t.expectLeases(descID, "/2/1")
 
 	// It is an error to acquire a lease for an old version once a new version
@@ -227,24 +292,32 @@ func TestLeaseManager(testingT *testing.T) {
 	l8 := t.mustAcquire(2, descID, 3)
 	t.expectLeases(descID, "/2/1 /2/2 /3/1 /3/2")
 
-	t.mustRelease(1, l5)
+	t.mustRelease(1, l5, &releaseWaiter)
 	t.expectLeases(descID, "/2/2 /3/1 /3/2")
-	t.mustRelease(2, l6)
+	t.mustRelease(2, l6, &releaseWaiter)
 	t.expectLeases(descID, "/3/1 /3/2")
 
 	// Wait for version 4 to be published.
 	wg.Wait()
 	l9 := t.mustAcquire(1, descID, 4)
-	t.mustRelease(1, l7)
-	t.mustRelease(2, l8)
+	t.mustRelease(1, l7, &releaseWaiter)
+	t.mustRelease(2, l8, nil)
 	t.expectLeases(descID, "/3/2 /4/1")
-	t.mustRelease(1, l9)
+	t.mustRelease(1, l9, nil)
 	t.expectLeases(descID, "/3/2 /4/1")
 }
 
 func TestLeaseManagerReacquire(testingT *testing.T) {
 	defer leaktest.AfterTest(testingT)()
 	ctx := server.MakeTestContext()
+	var releaseWaiter leaseReleaseWaiter
+	ctx.TestingKnobs = base.TestingKnobs{
+		SQLLeaseManager: &csql.LeaseManagerTestingKnobs{
+			LeaseStoreTestingKnobs: csql.LeaseStoreTestingKnobs{
+				LeaseReleasedEvent: releaseWaiter.LeaseReleasedNotification,
+			},
+		},
+	}
 	t := newLeaseTest(testingT, &ctx)
 	defer t.cleanup()
 
@@ -285,9 +358,9 @@ func TestLeaseManagerReacquire(testingT *testing.T) {
 	}
 	t.expectLeases(descID, "/1/1 /1/1")
 
-	t.mustRelease(1, l1)
-	t.mustRelease(1, l2)
-	t.mustRelease(1, l3)
+	t.mustRelease(1, l1, nil)
+	t.mustRelease(1, l2, &releaseWaiter)
+	t.mustRelease(1, l3, nil)
 }
 
 func TestLeaseManagerPublishVersionChanged(testingT *testing.T) {

